File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(timeLeft) {
  if (timeLeft === undefined) {
    return "You forgot to set the timer.";
  }
  if (timeLeft === 0) {
    return "Lasagna is done.";
  } else {
    return "Not done, please wait.";
  }
}

export function preparationTime(layers, timePerLayer = 2) {
  return layers.length * timePerLayer;
}

export function quantities(layers) {
  return {
    noodles: layers.filter(layer => layer === "noodles").length * 50,
    sauce: layers.filter(layer => layer === "sauce").length * 0.2,
  };
}

export function addSecretIngredient(friendList, myList) {
  myList.push(friendList[friendList.length - 1]);
}

export function scaleRecipe(originalRecipe, portions) {
  let newRecipe = {};

  for (const ingredient in originalRecipe) {
    newRecipe[ingredient] = (originalRecipe[ingredient] / 2) * portions;
  }
  return newRecipe;
}

File: go/lasagna-master/lasagna_master.go
package lasagna

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, preparationTime int) int {
	if preparationTime == 0 {
		preparationTime = 2
	}
	return len(layers) * preparationTime
}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (noodlesGrams int, sauceLiters float64) {
	countNoodles := 0
	countSauce := 0
	for _, layer := range layers {
		if layer == "noodles" {
			countNoodles++
		} else if layer == "sauce" {
			countSauce++
		}
	}
	return countNoodles * 50, float64(countSauce) * 0.2
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(friendsList, myList []string) {
	myList[len(myList)-1] = friendsList[len(friendsList)-1]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, portions int) []float64 {
	scale := float64(portions) / 2
	scaledQuantities := make([]float64, 0, len(quantities))
	for _, quantity := range quantities {
		scaledQuantities = append(scaledQuantities, quantity*scale)
	}
	return scaledQuantities
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
//
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more
// functionality.
